storage/namespace_readers.go

```diff
@@ -0,0 +1,313 @@
+	} else {
+		reader, err = m.newReaderFn(m.bytesPool, m.fsOpts)
+		if err != nil {
+			return nil, err
+		}
+
+		m.metrics.cacheMissAllocReader.Inc(1)
+	}
+	// Can release the lock now as we prepare reader for caller
+	m.Unlock()
+
+	if err := reader.Open(m.namespace.ID(), shard, blockStart); err != nil {
+		return nil, err
+	}
+	// We can validate metadata immediately since its read when opened
+	if err := reader.ValidateMetadata(); err != nil {
+		return nil, err
+	}
+
+	// Fast fwd through if in the middle of a volume
+	for i := 0; i < position.dataIdx; i++ {
+		id, data, _, err := reader.Read()
+		if err != nil {
+			return nil, err
+		}
+		id.Finalize()
+		data.Finalize()
+	}
+	for i := 0; i < position.metadataIdx; i++ {
+		id, _, _, err := reader.ReadMetadata()
+		if err != nil {
+			return nil, err
+		}
+		id.Finalize()
+	}
+
+	return reader, nil
+}
+
+func (m *namespaceReaderManager) put(reader fs.FileSetReader) {
+	status := reader.Status()
+
+	m.Lock()
+	defer m.Unlock()
+
+	if !status.Open {
+		m.closedReaders = append(m.closedReaders, cachedReader{
+			reader: reader,
+		})
+		return
+	}
+
+	key := cachedOpenReaderKey{
+		shard:      status.Shard,
+		blockStart: xtime.ToUnixNano(status.BlockStart),
+		position: readerPosition{
+			dataIdx:     reader.EntriesRead(),
+			metadataIdx: reader.MetadataRead(),
+		},
+	}
+
+	if _, ok := m.openReaders[key]; ok {
+		// Unlikely, however if so just close the reader we were trying to put
+		// and put into the closed readers
+		if err := reader.Close(); err != nil {
+			m.logger.Errorf("error closing reader on put from reader cache: %v", err)
+			return
+		}
+		m.closedReaders = append(m.closedReaders, cachedReader{
+			reader: reader,
+		})
+		return
+	}
+
+	m.openReaders[key] = cachedReader{
+		reader: reader,
+	}
+}
+
+func (m *namespaceReaderManager) tick() {
+	m.tickWithThreshold(expireCachedReadersAfterNumTicks)
+}
+
+func (m *namespaceReaderManager) close() {
+	// Perform a tick but make the threshold zero so all readers must be expired
+	m.tickWithThreshold(0)
+}
+
+func (m *namespaceReaderManager) tickWithThreshold(threshold int) {
+	m.Lock()
+	defer m.Unlock()
+
+	// First increment ticks since used for closed readers
+	expiredClosedReaders := 0
+	for i := range m.closedReaders {
+		m.closedReaders[i].ticksSinceUsed++
+		if m.closedReaders[i].ticksSinceUsed >= threshold {
+			expiredClosedReaders++
+		}
+	}
+	// Expire any closed readers, alloc a new slice to avoid spikes
+	// of use creating slices that are never released
+	if expired := expiredClosedReaders; expired > 0 {
+		newClosedReaders := make([]cachedReader, 0, len(m.closedReaders)-expired)
+		for _, elem := range m.closedReaders {
+			if elem.ticksSinceUsed < threshold {
+				newClosedReaders = append(newClosedReaders, elem)
+			}
+		}
+		m.closedReaders = newClosedReaders
+	}
+
+	// For open readers calculate and expire from map directly
+	for key, elem := range m.openReaders {
+		elem.ticksSinceUsed++
```

Comment:
the ++ is confusing because it makes it seem like you inadvertently introduced a bug by trying to mutate a copy. I can't think of a way to make the code more clear, so maybe a comment here saying "mutates the copy, but not the element in the map itself" and then on line 311 where you reassign "reassign the mutated copy"

Reply:
Sure thing.